### ISSUE TITLE
memory leak cased by function tf.dynamic_partition

### ISSUE THREAD

**Issue by mzhaoshuai** [tensorflow/tensorflow#22464]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:_no_
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  _Linux Ubuntu 16.04.4 LTS_。
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:_no_
- **TensorFlow installed from (source or binary)**:_binary_
- **TensorFlow version (use command below)**:_1.8.0_
- **Python version**:_Python 3.5.2_
- **Bazel version (if compiling from source)**:_None_
- **GCC/Compiler version (if compiling from source)**:_None_
- **CUDA/cuDNN version**:_CUDA-9.0/cuDNN-7.0.5_
- **GPU model and memory**: _GTX 1080Ti 11GB, 2 same cards_
- **Exact command to reproduce**:_below_

### Supplementary material
I just pulled the tensorflow's offical docker image and RUN some `pip install ...` instructions.
The tag of the iamge is `1.8.0-devel-gpu-py3`.

### Describe the problem
I want to calculate a focal loss. 
![image](https://user-images.githubusercontent.com/22476764/45917573-bb403600-beac-11e8-8262-56be63acd1c3.png)

It just added weight factor to the original cross entropy loss.

#### `tf_leak` code (shown in Source code / logs) 

When I used the `tf_leak` code as shown below, there will be memory leak.
I used `top` command to see the process's memory usage.

- **pic01 occupy about 3G memory after about 1min**
![image](https://user-images.githubusercontent.com/22476764/45917603-212cbd80-bead-11e8-8b2b-e784196405b2.png)

- **pic02 occupy about 11G memory after about 33min**
![image](https://user-images.githubusercontent.com/22476764/45917609-3ace0500-bead-11e8-9d6e-b58d748f6fbc.png)

- **pic03 occupy about 14G memory after about 191min**
![image](https://user-images.githubusercontent.com/22476764/45917619-45889a00-bead-11e8-8370-082b341c4907.png)

**I use the `tf.estimator` and the graph is finazied. So I do not add ops to the graph when the seesion is running.**

- **pic04 graph is finazied**
![image](https://user-images.githubusercontent.com/22476764/45917706-16bef380-beae-11e8-8af9-a0a8fd46c52a.png)

#### `tf_no_leak` code (shown in Source code / logs) 
 
I replace the ` tf.dynamic_partition` function with `tf.boolean_mask` and the memory leak disappered.

- **pic05 occupy about 3G memory after about 9min**
![image](https://user-images.githubusercontent.com/22476764/45917758-ec216a80-beae-11e8-9d49-115f2598fa50.png)

- **pic06 occupy about 3G memory after about 133min**
![image](https://user-images.githubusercontent.com/22476764/45917782-2be85200-beaf-11e8-9427-021f2f5fcd51.png)

#### more information

I tested the `tf_leak` and `tf_no_leak` code with a small dataset, when I use a bigger dataset, the process will occupy a large memory and the system will kill the process. This is why I found the problem.

### Source code / logs
```
params:
		valid_labels	:	the labels, a Tensor with shape [N * w * h, ]
		valid_logits	:	the logits, a Tensor with shape [N * w * h, num_classes]
		num_classes 	: 	the number of classes, a scalar
                ohem_prob_g  :       a hyperparameter, a scalar, default=1.5
```

1. `tf_leak` code

```	
valid_probs = tf.clip_by_value(tf.nn.softmax(valid_logits, axis=-1), \
					clip_value_min=1e-6, clip_value_max=1.0
valid_one_hot = tf.one_hot(valid_labels, depth=num_classes, dtype=tf.int32)
weight_matrix = ohem_prob_g - tf.dynamic_partition(valid_probs, valid_one_hot, num_partitions=2)[1]
## calculate the final cross-entropy
cross_entropy = tf.losses.sparse_softmax_cross_entropy(labels=valid_labels, 
						logits=valid_logits, weights=weight_matrix)
return cross_entropy
```

2. `tf_no_leak` code

same as the `tf_leak` code except I replace the `tf.dynamic_partition` with `tf.boolean_mask`.

```
valid_probs = tf.clip_by_value(tf.nn.softmax(valid_logits, axis=-1), \
					clip_value_min=1e-6, clip_value_max=1.0)
valid_one_hot = tf.one_hot(valid_labels, depth=num_classes, dtype=tf.int32)
## calculate the final cross-entropy
weight_matrix = ohem_prob_g - tf.boolean_mask(valid_probs, valid_one_hot)
cross_entropy = tf.losses.sparse_softmax_cross_entropy(labels=valid_labels, 
						logits=valid_logits, weights=weight_matrix)
return cross_entropy
```

3. input code used for the `tf.estimator`

```
def input_fn(is_training, data_dir, batch_size, num_epochs=1):
  """Input_fn using the tf.data input pipeline for CIFAR-10 dataset.
  Args:
    is_training: A boolean denoting whether the input is for training.
    data_dir: The directory containing the input data.
    batch_size: The number of samples per batch.
    num_epochs: The number of epochs to repeat the dataset.
  Returns:
    A tuple of images and labels.
  """
  dataset = tf.data.Dataset.from_tensor_slices(get_filenames(is_training, data_dir))
  dataset = dataset.flat_map(tf.data.TFRecordDataset)
  if is_training:
    dataset = dataset.shuffle(buffer_size=_NUM_IMAGES['train'])

  dataset = dataset.map(parse_record)
  dataset = dataset.map(
      lambda image, label, idx: preprocess_image(image, label, idx, is_training),
      num_parallel_calls=1)

  # We call repeat after shuffling, rather than before, to prevent separate
  # epochs from blending together.
  dataset = dataset.repeat(num_epochs)
  dataset = dataset.batch(batch_size)
  dataset = dataset.prefetch(buffer_size=FLAGS.prefetch_buffer_size)

  iterator = dataset.make_one_shot_iterator()
  images, labels, idx = iterator.get_next()

  return images, labels

```

**Comment by mzhaoshuai** [comment]
#### Supplementary information
When I use the results of the `tf.dynamic_partition` to do comparison operations(`>,<,==`) with a scalar, it is totally ok.

**Comment by jiarenyf** [comment]
I meet similar question, under watching.

**Comment by mzhaoshuai** [comment]
Did anybody else meet this question?

**Comment by harshini-gadige** [comment]
@rmlarsen Hi, could you please look into this issue ?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sushreebarsa** [comment]
@mzhaoshuai Is this still an issue?
We see that you are using TF v1.x which is not actively supported and we recommend you to upgrade latest Tf versions(2.4 or later).Please refer to the [migration](https://www.tensorflow.org/guide/migrate) doc .
Estimators are not recommended for new code. Estimators  can behave unexpectedly, especially when combined with TF 2 code. Could you please refer to this [link](https://www.tensorflow.org/api_docs/python/tf/estimator/DNNClassifier) for more details  and let us know if it helps?
Thanks!
